jest-common/src/main/java/io/searchbox/client/AbstractJestClient.java

```diff
@@ -115,13 +115,53 @@ public ServerPool(final Set<String> servers) {
         }
 
         public String getNextServer() {
-            if (serverIterator.hasNext()) return serverIterator.next();
+            String server = serversRing.next();
+            if(server != null) return server;
             else throw new NoServerConfiguredException("No Server is assigned to client to connect");
         }
 
         public int getSize() {
             return servers.size();
         }
+    }
+
+    private static final class ImmutableCircularList<E> {
+
+        final int size;
+        Node<E> current;
 
+        private static final class Node<E> {
+            E item;
+            Node<E> next;
+
+            Node(E data, Node<E> next) {
+                this.item = data;
+                this.next = next;
+            }
+        }
+
+        public ImmutableCircularList(Collection<? extends E> c) {
+            size = c.size();
+            if (size == 0)
+                return;
+
+            Iterator<?> iter = c.iterator();
+            Node<E> head = new Node<E>((E) iter.next(), null);
+            current = head;
+            while (iter.hasNext()) {
+                Node<E> nextNode = new Node<E>((E) iter.next(), null);
+                current.next = nextNode;
+                current = nextNode;
+            }
+            current.next = head;
+            current = head;
+        }
+
+        public E next() {
+            if(current == null) return null;
+            E data = current.item;
```

Comment:
I think this is as fast as you are going to get without a side effect https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/atomic/package-summary.html

You must have some performance measurement perhaps you could share

BTW, This is the "implementation" of `Iterators.cycle()` that is causing the issue.  It has the same issue as this code, it just manifests itself in a different way. The iterator is exhausted at `!iterator.hasNext())` but one thread is calling `iterator.next()` before the assignment of `iterator = iterable.iterator();` which restarts the cycle.

      @Override
      public boolean hasNext() {
        if (!iterator.hasNext()) {
          iterator = iterable.iterator();
        }
        return iterator.hasNext();
      }